### PR TITLE
Implement schema support

### PR DESCRIPTION
**Pull request by jeroenvandisseldorp**
Adds basic schema support, plus fixes many smaller issues.

### CODE REVIEW

File: examples/test.avsc [hunk 1]
```diff
@@ -0,0 +1,68 @@
+{
```
Comment (richard-axual) [inline examples/test.avsc:1]: This file is named test.
If this is part of the examples it should be renamed. Or else it should be removed

**Review by richard-axual — CHANGES_REQUESTED** [review]

**Comment by tonvanbart** [comment]
There are 242 changed files in this PR, which makes it a bit unwieldy to review. Some (most?) of the changes appear to be non related to schema support, or is that just me?
It would be better if we split this PR into two, one for the formatting/small issues changes and one for the new functionality.

File: ksml/src/main/java/io/axual/ksml/data/mapper/UserTypeStringMapper.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        if (value instanceof ListType listType) return listTypeToString(listType);
+        if (value instanceof StructType structType) return structTypeToString(structType);
+        if (value instanceof TupleType tupleType) return tupleTypeToString(tupleType);
+        return null;
+    }
+
+    private String enumTypeToString(EnumType enumType) {
+        var result = new StringBuilder("enum(");
+        var first = true;
+        for (String symbol : enumType.symbols()) {
```
Comment (tonvanbart) [inline ksml/src/main/java/io/axual/ksml/data/mapper/UserTypeStringMapper.java:71]: Could be made simpler using streams and Collectors.joining()

File: ksml/src/main/java/io/axual/ksml/data/mapper/UserTypeStringMapper.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    @Override
+    public String toString(UserType value) {
+        var notation = value.notation().toLowerCase();
+        var type = toString(value.dataType());
+        if (type == null || type.isEmpty()) return notation;
+        return notation + ":" + type;
+    }
+
+    private String toString(DataType value) {
+        if (value == DataNull.DATATYPE) return "null";
```
Comment (tonvanbart) [inline ksml/src/main/java/io/axual/ksml/data/mapper/UserTypeStringMapper.java:51]: Might be better to move this into some toString() method inside the enum

**Review by tonvanbart — CHANGES_REQUESTED** [review]
None of the added files have any test coverage, I think there should be some test code at least.

File: examples/test.avsc (hunk 1 above)
Comment (jeroenvandisseldorp) [reply examples/test.avsc]: Removed

File: examples/test1.json [hunk 4]
```diff
@@ -0,0 +1,7 @@
+{"schema":"{"type":"record","name":"SensorData","namespace":"io.axual.ksml.example","doc":"","fields":[{"name":"name","type":"string","doc":""},{"name":"timestamp","type":"long","doc":""},{"name":"value","type":"string","doc":""},{"name":"type","type":{"type":"enum","name":"SensorType","doc":"","symbols":["AREA","HUMIDITY","LENGTH","STATE","TEMPERATURE"]},"doc":""},{"name":"unit","type":"string","doc":""},{"name":"color","type":["null","string"],"doc":"","default":null},{"name":"city","type":["null","string"],"doc":"","default":null},{"name":"owner","type":["null","string"],"doc":"","default":null}]}"}
```
[the remark this answers is not in the document]
Comment (jeroenvandisseldorp) [reply examples/test1.json]: Removed

File: ksml/src/main/java/io/axual/ksml/data/mapper/UserTypeStringMapper.java (hunk 3 above)
Comment (jeroenvandisseldorp) [reply ksml/src/main/java/io/axual/ksml/data/mapper/UserTypeStringMapper.java:51]: I considered this, but the mapper goes two-way, namely from/to the formal syntax of the UserType. Implementing this in the class itself would mean that the class itself knows its formal syntax. Normally toString() in a class is only used to aid developers or for one-way conversion to String, but not to be parsed back.

**Comment by jeroenvandisseldorp** [comment]
> There are 242 changed files in this PR, which makes it a bit unwieldy to review. Some (most?) of the changes appear to be non related to schema support, or is that just me? It would be better if we split this PR into two, one for the formatting/small issues changes and one for the new functionality.

Valid point, and discussed with Richard as well last week, but not something we can do much about at this stage.

File: ksml/src/main/java/io/axual/ksml/data/mapper/UserTypeStringMapper.java (hunk 2 above)
Comment (jeroenvandisseldorp) [reply ksml/src/main/java/io/axual/ksml/data/mapper/UserTypeStringMapper.java]: Done

**Review by richard-axual — APPROVED** [review]
Approved, but we really need to start working on tests and refactoring to make KSML easy to test

**Pull request merged**